### PR TITLE
ATM: Simplify query configurations

### PR DESCRIPTION
**Pull request by tiferet**
This PR simplifies the query configurations. It moves common functionality from the derived classes up to `AtmConfig`. It also combines the two configs each query previously had into a single config, simplifying the code and making it easier to understand. This will prevent code duplication and bugs, improve maintainability, and make it easier to add new queries.

Once again commit-by-commit review is recommended 😄 

### Main changes

- `AtmConfig` inherits from `TaintTracking::Configuration`. That way the specific query configs (e.g. `DomBasedXssAtmConfig`), which inherit from `AtmConfig`, also inherit from `TaintTracking::Configuration`. This removes the need for two separate config classes for each query (e.g. `DomBasedXssAtmConfig` and `XssATM::Configuration`). Note: We could also keep `AtmConfig` as is and have the query configs inherit both from it and from `TaintTracking::Configuration` directly, but I think having `AtmConfig` inherit from `TaintTracking::Configuration` makes more sense: it's a query configuration for a boosted taint tracking query, which is a type of taint tracking query that adds a few extra predicates (e.g. `isKnownSink`).
- Move the definition of `isSource` to the base class: A long as we're not boosting sources, `isSource` is identical to `isKnownSource`.
- Move the definition of `isSink` to the base class: Holds if `sink` is a known taint sink or an "effective" sink (a candidate to be classified by an ML model).
- Remove code duplication in query .ql files: Define the query for finding ATM alerts in the base class `AtmConfig`, and call it from each query's .ql file.

### Consistency check

[This DCA experiment](https://github.com/github/codeql-dca-main/issues/8683) uses the model on `main` with the final commit of each PR from the CodeQL revamp. The idea is to double-check that none of the changes I've made has affected the endpoints that get scored at inference time, since this is not directly tested for in the PR checks. We see identical results from all variants (e.g. [this](https://github.com/github/codeql-dca-main/blob/data/tiferet/codeql-refactoring3/reports/summaries/csv/Alert_count_using_manual_result_classifications_per_source_and_query_(v0_vs_v4).csv)), confirming that none of the revamp work has changed the endpoints being scored.

### Timing checks
- ✅ KPI timing experiment: https://github.com/github/codeql-dca-main/issues/8684
- ☑️ The local runtime of `endpoint_large_scale/ExtractEndpointDataTraining` remains like it was after the [last PR that affected timing](https://github.com/github/codeql/pull/11263): About 5s.

Addresses most of https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2134.

### CODE REVIEW

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll
Comment (tiferet) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll:68]: I don't know why in this one case `isSource` is not identical to `isKnownSource`. If this was done on purpose, I can override `isSource` in this one query with the original definition it had here. Removing it didn't fail the tests, though, and seems to make logical sense: I find it surprising that `isSource` should ever include _fewer_ endpoints than `isKnownSource`. Please let me know if there's a reason for this difference between `isSource` and `isKnownSource`, and I'll fix this 🙏

**Comment by tiferet** [comment]
@kaeluka when you review this PR, do you mind looking at the possible issue Henry raised [here](https://github.slack.com/archives/CP9127VUK/p1669045505695319). I _think_ it's OK, but I have limited understanding of the issue he's pointing out. Henry said `From a very very quick scan, it looks OK, but I’d recommend taking a closer look.` I'd appreciate your eyes on it as well to make sure there isn't a problem.

**Comment by kaeluka** [comment]
Thank you for point this out! I was going to, but I *definitely* want to talk to Henry about it as well!

**Comment by kaeluka** [comment]
As part of digging into the questions brought up by Henry, I added the following snippet to each file modified in this PR and quick-evaluated it:

```ql
predicate test(DataFlow::Configuration a) {
  any()
}
```

This showed me that there's actually a few files where there's several instances of `DataFlow::Configuration` in scope. I believe that, in each case, the situation is not preventing the PR from being merged.

These files are:
 - :white_check_mark: `EndpointFeatures.ql` - benign because the query only computes features, rather than evaluating flow.
 - :white_check_mark: `FilteredTruePositives.ql` - benign because the query only runs queries involving characteristics, rather than evaluating flow (and it's just a unit test).
 - :white_check_mark: `ExtractEndpointMapping.ql` - benign because the query does not evaluate flow
 - :white_check_mark: `DebugResultInclusion.ql` - this might actually be slowed down because it uses `hasFlow` *but* a) it's only a debug query, and b) the situation is like that already on `main` (the PR only changes some names in the file).

@henrymercer, is my reasoning is sound?

**Comment by henrymercer** [comment]
> @henrymercer, is my reasoning sound?

Your reasoning around the files you listed looks good.  Most of those files are test or debug files.  I'm not sure how you handled the case of libraries being changed, as we'd need to check all the files that use this library and not just the ones being changed.

I did try evaluating the predicate top-down on each of the queries and they each reported a single dataflow configuration, so I think we're good to go.

**Comment by tiferet** [comment]
@kaeluka Since you're already in the middle of reviewing this PR, I don't want to rebase it on top of the latest version of `tiferet/complexity-reduction` and interfere with your ongoing work. Please LMK what your preference is. Should I:
- Rebase on top of `tiferet/complexity-reduction` now?
- Wait until `tiferet/complexity-reduction` is merged to main and then rebase on top of main?
- Wait until `tiferet/complexity-reduction` is merged to main and then merge main into this branch?
- Something else?

File: javascript/ql/experimental/adaptivethreatmodeling/src/NosqlInjectionATM.ql
Comment (kaeluka) [inline javascript/ql/experimental/adaptivethreatmodeling/src/NosqlInjectionATM.ql:21]: 👍: the query certainly reads nicer now and the logic is now confined to the .qll files. I like it.

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/EndpointFeatures.ql
Comment (kaeluka) [inline javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/EndpointFeatures.ql:23]: 🙃: These changes are only due to the somewhat (IMHO) needless renaming. Last time I'll bring it up :)

**Review by kaeluka — APPROVED** [review]
The big question, whether this might introduce a pathological dependency graph, here has been tackled in the PR discussion already.

👍: The rest of the review is that I:
 - Like the change and am not asking you to make any big changes.
 - I'd slightly prefer to follow convention in the naming of the Config classes but it's not a hill I'll die on.
 - I was a bit surprised to see the PR, as it seems like it's a bit of a side track from the grand design we're following. Happy to see it either way, though!

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll:14]: I tried to follow the convention on configurations as I understood it:  The existing code was
```
class Configuration extends TaintTracking::Configuration {
  Configuration() { this = "NosqlInjectionATM" }
```
One level up, in TaintTracking.qll, it also reads
```
abstract class Configuration extends DataFlow::Configuration
```
I understood this to mean that for some reason the convention is that configuration classes are always called `Configuration`, rather than something more specific such as `NosqlInjectionATMConfig`. I don't particularly like this either, so if there's no reason to do it this way I'm happy to change it!

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll]: Good call -- fixed in 6813358037124d023d313fbc135d2f09719720d4. It looks like there were plans to do more in `isFlowLikelyInBaseQuery` that were never implemented:
```
     * Indicates whether the flow from source to sink is likely to be reported by the base security
     * query.
     *
     * Currently this is a heuristic: it ignores potential differences in the definitions of
     * additional flow steps.
```

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll:14]: @adityasharad in the interest of getting an answer today, do you know whether such a convention exists? If not, I'll make the suggested change and push a commit so @kaeluka can re-approve the PR tomorrow.

Comment (adityasharad) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll:14]: I don't think it's a strict convention; you see both in the libraries across languages. I think it's fine to keep this diff smaller and keep the string and class names in sync.

Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll:14]: Perfect -- thank you!

Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/NosqlInjectionATM.qll]: Fixed in c5184d37e7c8b3cfafaf5dcafab850589315bfac -- I do think this convention is nicer! Note that I changed e.g. `NosqlInjectionATMConfig` to `NosqlInjectionAtmConfig` to fit the style guide that acronyms should have only the first letter capitalized (e.g. `Http` rather than `HTTP`).

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (adityasharad) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:150]: Minor/bikeshed: I'd prefer a name that doesn't suggest a return/result value, and mentions flow. Also makes it easier to see the parallels between unboosted and boosted queries. How about `hasLikelyFlowPath`?

**Comment by tiferet** [comment]
> I was a bit surprised to see the PR, as it seems like it's a bit of a side track from the grand design we're following. Happy to see it either way, though!

Yes, I think we're already done implementing the design. All the issues left [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/1898#issue-1269359426) are improvements we can make either to the code or to the data now that the refactoring is complete 🎉

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:150]: I don't know about `hasLikelyFlowPath`, because one of the conditions it _must_ meet is `hasFlowPath`. Logically `hasFlowPath` should imply `hasLikelyFlowPath`. 

Maybe something like `hasBoostedFlowPath`? We've been trying to get rid of the "boosted" terminology, but I can't think of something better right now. I think `hasAlert` would be clearer (ultimately that's what this predicate does -- it surfaces all the alerts for this config), but that doesn't mention flow.

Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:150]: Note: I tracked this renaming together with the other renamings [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2136), so that we can merge this PR and XssThroughDOM and open a separate PR for all the renamings

**Review by adityasharad — APPROVED** [review]

**Pull request merged**
